base/operators.jl

```diff
@@ -1060,7 +1074,8 @@ julia> filter(!isletter, str)
 "∀  > 0, ∃  > 0: |-| <  ⇒ |()-()| < "
 ```
 """
-!(f::Function) = (x...)->!f(x...)
+!(f::Function) = (!) ∘ f
+!(f::ComposedFunction{typeof(!)}) = f.inner #allows !!f === f
```

Comment:
This may come as a surprise if there is a `!` definition for some type `T` such that `!!(x::T) != x`, as then we may have `!!f(x) != (!!f)(x)` if the return type of `f` is `T`. But probably that's just an argument for not defining such a `!`.

Reply:
I was wondering about this. It wouldn't cause problems to allow `!!f` definitions even though I can't imagine a reason to do so.

Reply:
That's a good point, @martinholters, but I think your take that it's an argument for not doing that is where I fall on that. If they really want to they can always overload `!` for their function as an escape hatch.